crates/storage/src/connection/event.rs

```diff
@@ -145,455 +105,159 @@ pub(super) fn get_events<K: KeyFilter>(
-                    param: (":events_match", key_fts_expression.to_sql()),
-                })
-            }
-        }
-    }
-}
-
-fn event_query<'query, 'arg>(
-    base: &'query str,
-    from_block: Option<&'arg BlockNumber>,
-    to_block: Option<&'arg BlockNumber>,
-    contract_address: Option<&'arg ContractAddress>,
-    keys: &'arg (dyn KeyFilter + 'arg),
-    strategy: QueryStrategy,
-) -> (
-    std::borrow::Cow<'query, str>,
-    Vec<(&'static str, rusqlite::types::ToSqlOutput<'arg>)>,
-) {
-    let mut base_query = std::borrow::Cow::Borrowed(base);
-
-    let mut where_statement_parts: Vec<&'static str> = Vec::new();
-    let mut params: Vec<(&str, rusqlite::types::ToSqlOutput<'arg>)> = Vec::new();
-
-    // filter on block range
-    match (from_block, to_block) {
-        (Some(from_block), Some(to_block)) => {
-            where_statement_parts.push("block_number BETWEEN :from_block AND :to_block");
-            params.push((":from_block", from_block.to_sql()));
-            params.push((":to_block", to_block.to_sql()));
-        }
-        (Some(from_block), None) => {
-            where_statement_parts.push("block_number >= :from_block");
-            params.push((":from_block", from_block.to_sql()));
-        }
-        (None, Some(to_block)) => {
-            where_statement_parts.push("block_number <= :to_block");
-            params.push((":to_block", to_block.to_sql()));
+                tracing::trace!("Bloom filter matched");
+            };
         }
-        (None, None) => {}
-    }
 
-    // on contract address
-    if let Some(contract_address) = contract_address {
-        where_statement_parts.push("from_address = :contract_address");
-        params.push((":contract_address", contract_address.to_sql()));
-    }
-
-    // Filter on keys: this is using an FTS5 full-text index (virtual table) on the keys.
-    // The idea is that we convert keys to a space-separated list of Base64/Base32 encoded string
-    // representation and then use the full-text index to find events matching the events.
-    if let Some(result) = keys.apply(strategy) {
-        base_query.to_mut().push_str(result.base_query);
-        where_statement_parts.push(result.where_statement);
-        params.push((result.param.0, result.param.1));
-    }
-
-    if !where_statement_parts.is_empty() {
-        let needed = " WHERE ".len()
-            + where_statement_parts.len() * " AND ".len()
-            + where_statement_parts.iter().map(|x| x.len()).sum::<usize>();
+        let block_header = tx.block_header(crate::BlockId::Number(block_number))?;
+        let Some(block_header) = block_header else {
+            break;
+        };
 
-        let q = base_query.to_mut();
-        if let Some(more) = needed.checked_sub(q.capacity() - q.len()) {
-            q.reserve(more);
+        blocks_scanned += 1;
+        if blocks_scanned > MAX_BLOCKS_TO_SCAN {
+            return Err(EventFilterError::TooManyMatches);
```

Comment:
Should this not return an empty set with the continuation token incremented?

Reply:
It should (although not an empty set but the current event set).

It took me quite some time to implement this because our previous split between the storage & RPC code made this impossible. Now the storage code handles continuation tokens so that it can actually return a partial page with a continuation token in case we reach the block scan limit.

We also had an assumption that a continuation token can be returned if and only if we have a full page -- that's also not true anymore.

I'd like to make the block scan limit configurable so that operators can set it for themselves (basically means: how long a `starknet_getEvents` call can take).

Maybe we could also introduce a limit on the number of (non-cached) Bloom lookups.